### PR TITLE
feat: Add optimization-type–aware parallelization and picking to AIC

### PR DESCRIPTION
**Pull request by ashnamehrotra**
## feat: Add optimization-type–aware parallelization and picking to AIC

Companion PR to dynamo#6134 (Make SLAs optional with dynamic default presets).

Adds MoE/MLA-aware parallelization to the naive generator and a new `pick_optimization_type()` picking mode, enabling the Dynamo profiler to request configs optimized for throughput or latency without explicit SLA targets.

#### `generator/naive.py`
- New `_resolve_parallelization()` helper that selects the parallelization strategy based on model architecture and optimization objective:
  - **Dense models** → TP (tensor parallelism)
  - **MoE models** → TEP (tensor-expert parallelism via `moe_tensor_parallel_size`)
  - **MLA+MoE + throughput** (DeepSeek-V3/V2) → DEP (data-expert parallelism via `data_parallel_size` + `moe_expert_parallel_size`)
- `build_naive_generator_params()` now accepts `mode` (`"agg"` / `"disagg"`) and `optimization_type` (`"throughput"` / `"latency"` / `None`) parameters
- Detects MoE architecture automatically from model config
- Disaggregated mode support: generates separate `prefill` and `decode` worker params
- Fully backward-compatible — existing callers without the new args get identical behavior

#### `sdk/picking.py`
- New `pick_optimization_type(pareto_df, optimization_type, total_gpus, serving_mode, top_n)` function
  - **throughput**: ranks by `tokens/s/gpu_cluster` descending (tiebreak: lower tpot)
  - **latency**: ranks by `tpot` ascending (tiebreak: higher throughput)
  - No SLA filtering — purely objective-driven ranking
  - Returns standard `{best_config_df, best_throughput, best_latencies, pareto_frontier_df}` dict
- Follows existing patterns (`pick_default`, `pick_load_match`, `pick_autoscale`)

### Tests

- `tests/unit/generator/test_naive_optimization_type.py` — 14 tests covering `_resolve_parallelization` (dense TP, MoE TEP, MLA+MoE DEP, edge cases) and `build_naive_generator_params` (agg/disagg modes, parallelization keys, worker config structure)
- `tests/unit/sdk/test_pick_optimization_type.py` — 11 tests covering throughput/latency ranking, tiebreaking, dedup by parallel key, empty/None DataFrames, top_n limiting, pareto frontier

All 50 unit tests pass (25 existing naive + 25 new).

## Summary by CodeRabbit

* **New Features**
  * Added intelligent support for mixture-of-experts (MoE) model architectures in parallelization strategy selection.
  * Introduced optimization-type aware configuration picking: users can now optimize for throughput or latency preference.
  * Added disaggregated serving mode for separate prefill/decode parameter configuration alongside aggregated mode.

### CODE REVIEW

File: src/aiconfigurator/generator/naive.py [hunk 1]
```diff
@@ -212,33 +274,31 @@ def build_naive_generator_params(
...
-                    "pipeline_parallel_size": int,
-                    "max_batch_size": int,
-                    ...
-                }
-            }
-        }
+        Dictionary containing generator parameters.  When ``mode="agg"``,
+        ``params.agg`` is populated.  When ``mode="disagg"``, both
+        ``params.prefill`` and ``params.decode`` are populated with
+        identical parallelization.
```
Comment (tedzhouhk) [inline src/aiconfigurator/generator/naive.py:302]: we can consider a better default bs and mnt for prefill, but can be in another PR, maybe add a TODO?

File: src/aiconfigurator/generator/naive.py [hunk 2]
```diff
@@ -248,63 +308,133 @@ def build_naive_generator_params(
...
         gpus_per_node=gpus_per_node,
         total_gpus=total_gpus,
     )
-    pipeline_parallel_size = 1
+
+    # Detect model architecture for MoE-aware parallelization
+    architecture = ""
+    is_moe = False
+    try:
+        from aiconfigurator.sdk.utils import get_model_config_from_model_path
```
Comment (tedzhouhk) [inline src/aiconfigurator/generator/naive.py]: import at top?

File: src/aiconfigurator/sdk/picking.py [hunk 3]
```diff
@@ -476,3 +476,117 @@ def pick_autoscale(
...
+    Returns:
+        Dict with keys:
+        - ``best_config_df``: Top-N configurations DataFrame, sorted by
+          the optimization objective.
+        - ``best_throughput``: ``tokens/s/gpu_cluster`` of the rank-1 config.
+        - ``best_latencies``: Dict with ``ttft``, ``tpot``,
+          ``request_latency`` from the rank-1 config.
+        - ``pareto_frontier_df``: Pareto frontier DataFrame.
+    """
+    from aiconfigurator.sdk.pareto_analysis import get_pareto_front
```
Comment (tedzhouhk) [inline src/aiconfigurator/sdk/picking.py:524]: import at top?

File: src/aiconfigurator/generator/naive.py [hunk 4]
```diff
@@ -248,63 +308,133 @@ def build_naive_generator_params(
...
+            "NodeConfig": {
+                "num_gpus_per_node": gpus_per_node,
+            },
+            "WorkerConfig": {
+                "prefill_workers": prefill_workers,
+                "prefill_gpus_per_worker": min_gpus,
+                "decode_workers": decode_workers,
+                "decode_gpus_per_worker": min_gpus,
+            },
+            "backend": backend_name,
```
[the remark this answers is not in the document]
Comment (tedzhouhk) [reply src/aiconfigurator/generator/naive.py:417]: +1, maybe add a warning here?

File: src/aiconfigurator/generator/naive.py [hunk 5]
```diff
@@ -248,63 +308,133 @@ def build_naive_generator_params(
...
+        }
```
[the remark this answers is not in the document]
Comment (tedzhouhk) [reply src/aiconfigurator/generator/naive.py:418]: +1

**Review by tedzhouhk — APPROVED** [review]

File: src/aiconfigurator/generator/rule_plugin/vllm.rule [hunk 6]
```diff
@@ -2,6 +2,13 @@ BenchConfig.estimated_concurrency = (BenchConfig.estimated_concurrency if BenchC
 agg_decode max_batch_size = (512 if (max_batch_size or 0) < 512 else (max_batch_size * 2))
 prefill max_batch_size = (max_batch_size if max_batch_size else 1)
 
+# Enforce vLLM MoE parallelism: vLLM uses unified TP for both attention and MoE,
+# so tensor_parallel_size must account for moe_tp × moe_ep.
+# Without this, MoE models get tp=1 and the full model is replicated per GPU, causing OOM.
+when ModelConfig.is_moe and (moe_tensor_parallel_size and moe_expert_parallel_size):
+    agg_prefill_decode tensor_parallel_size = moe_tensor_parallel_size * moe_expert_parallel_size
```
Comment (simone-chen) [inline src/aiconfigurator/generator/rule_plugin/vllm.rule]: @coderabbitai I think this is only true if dp=1.
In `utils.py::enumerate_parallel_config` we only add a combo to the parallel list if `dp * tp == moe_tp * moe_ep`. Should the correct logic here be:
`agg_prefill_decode tensor_parallel_size = moe_tensor_parallel_size * moe_expert_parallel_size / (data_parallel_size or 1)`

File: src/aiconfigurator/generator/api.py
Comment (simone-chen) [inline src/aiconfigurator/generator/api.py:1]: @coderabbitai Could you please help me confirm:
Line 637-639 and line 675-677 seem to be replicated. 
Python is function scoped so even if 637-639 is wrapped in a if condition, the variables should be accessible later at line 675 right?

File: src/aiconfigurator/generator/api.py [hunk 7]
```diff
@@ -634,8 +634,9 @@ def generate_naive_config(
         import random
 
         model_name_safe = model_path.replace("/", "_")
-        tp = generator_params["params"]["agg"]["tensor_parallel_size"]
-        pp = generator_params["params"]["agg"]["pipeline_parallel_size"]
+        first_params = generator_params["params"].get("agg") or generator_params["params"].get("prefill", {})
+        tp = first_params.get("tensor_parallel_size", 1)
+        pp = first_params.get("pipeline_parallel_size", 1)
```
Comment (simone-chen) [inline src/aiconfigurator/generator/api.py]: NIT: Shall we hoist these 3 lines before the `if` condition to avoid code duplication with line 675 to 677?
Would also prefer to rename `first_params` to `agg_or_prefill_params` 🙏

File: src/aiconfigurator/sdk/picking.py [hunk 8]
```diff
@@ -476,3 +476,117 @@ def pick_autoscale(
...
+
+
+# ---------------------------------------------------------------------------
+# pick_optimization_type
+# ---------------------------------------------------------------------------
+
+
+def pick_optimization_type(
+    pareto_df: pd.DataFrame,
+    optimization_type: str,
```
Comment (simone-chen) [inline src/aiconfigurator/sdk/picking.py]:
```suggestion
    optimization_type: Literal["throughput", "latency"],
```
Also needs to add import `from typing import Literal`

File: tests/unit/sdk/picking/test_pick_optimization_type.py
Comment (simone-chen) [inline tests/unit/sdk/picking/test_pick_optimization_type.py:1]: Thanks for adding tests! Do you mind creating a `picking` subfolder under `tests/unit/sdk` and move this test file there? Ideally we should also add test files like this for other picking strategies. Adding a `picking` folder will make it easier in future.

File: tests/unit/generator/test_naive_optimization_type.py [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+# build_naive_generator_params — mode and optimization_type
+# ---------------------------------------------------------------------------
+
+
+@pytest.mark.unit
+class TestBuildNaiveOptimizationType:
+    """Test build_naive_generator_params with mode and optimization_type args."""
+
+    _SYS: ClassVar[dict] = {"gpus_per_node": 8, "vram_per_gpu": 141 * 1024**3}
+    _WEIGHT = 30 * 1024**3  # 30 GiB -> fits in 1 GPU
```
Comment (simone-chen) [inline tests/unit/generator/test_naive_optimization_type.py:124]: @coderabbitai Do you think it's worth adding another test case where the model weight can't fit into 1 GPU to test smart TP selection?

**Review by simone-chen — APPROVED** [review]
Thanks for the PR. Left a few comments.

File: tests/unit/generator/test_naive_optimization_type.py (hunk 9 above)
Comment (ashnamehrotra) [reply tests/unit/generator/test_naive_optimization_type.py:124]: added
